File: benchmark-tablets/pipette-clients/ios/Pipette/Pipette/PlanTypes/BenchmarkDefinition.swift
import Foundation

/// A benchmark definition decoded from the server-shaped payload — the typed
/// Swift mirror of the Rust `BenchmarkDefinition` (`pipette_plan_types::benchmark`),
/// a sum type tagged on `benchmark_type` with each variant's typed parameters.
///
/// This replaces the per-runtime minimal JSON decoders (`MLXRuntime.BenchDef` /
/// `EvalDef` and `HeadlessRunner.Def`): the JSON is decoded once at the boundary
/// and the runtime dispatches on the case, so params are typed and there's no
/// re-parsing per kernel. Decoding rejects an unknown `benchmark_type` — the
/// client operates strictly internally (see PIP-235's loose→strict principle).
nonisolated enum BenchmarkDefinition: Equatable {
    case prefillThroughput(benchmarkId: String, prefillTokens: UInt32)
    case decodeThroughput(benchmarkId: String, prefillTokens: UInt32, decodeTokens: UInt32)
    case endToEndLatency(benchmarkId: String, prefillTokens: UInt32, decodeTokens: UInt32)
    case maxMemoryUsage(benchmarkId: String, prefillTokens: UInt32)
    case eval(
        benchmarkId: String, evalId: EvalId, datasetName: String,
        maxTokens: UInt32, mcqChoices: [String]?, samples: [EvalSample]?)
    case vlThroughput(
        benchmarkId: String, imageWidth: UInt32, imageHeight: UInt32,
        textTokens: UInt32, decodeTokens: UInt32)

    var benchmarkId: String {
        switch self {
        case .prefillThroughput(let id, _), .maxMemoryUsage(let id, _),
            .decodeThroughput(let id, _, _), .endToEndLatency(let id, _, _),
            .eval(let id, _, _, _, _, _), .vlThroughput(let id, _, _, _, _):
            return id
        }
    }

    /// Reconstruct a definition from a structured benchmark id, for the four
    /// ladder types — the execution path's fallback when the synced catalog has
    /// no entry for an id (e.g. a historical job whose benchmark left the
    /// catalog). Mirrors the Rust `BenchmarkType::from_id`: the type is the
    /// matched prefix and the workload numbers are parsed straight out of the id,
    /// so no ladder of values is hardcoded.
    ///
    /// - `prefill_throughput_<P>`    → `.prefillThroughput(id, P)`
    /// - `max_memory_usage_<P>`      → `.maxMemoryUsage(id, P)`
    /// - `decode_throughput_<P>_<D>` → `.decodeThroughput(id, P, D)`
    /// - `end_to_end_latency_<P>_<D>`→ `.endToEndLatency(id, P, D)`
    ///
    /// Returns `nil` for `eval`, `vl_throughput`, smoke, or anything that doesn't
    /// match a known type with the expected trailing integers. iOS benchmark ids
    /// here, so none is stripped.
    init?(parsingId id: String) {
        // Match on the type rawValue prefix (types contain underscores, so we
        // can't just split on `_`), then parse the trailing `_<P>` / `_<P>_<D>`.
        func suffixInts(after type: String) -> [UInt32]? {
            guard id.hasPrefix(type + "_") else { return nil }
            let tail = id.dropFirst(type.count + 1)
            let parts = tail.split(separator: "_", omittingEmptySubsequences: false)
            let nums = parts.compactMap { UInt32($0) }
            return nums.count == parts.count ? nums : nil
        }

        if let n = suffixInts(after: "prefill_throughput"), n.count == 1 {
            self = .prefillThroughput(benchmarkId: id, prefillTokens: n[0])
        } else if let n = suffixInts(after: "max_memory_usage"), n.count == 1 {
            self = .maxMemoryUsage(benchmarkId: id, prefillTokens: n[0])
        } else if let n = suffixInts(after: "decode_throughput"), n.count == 2 {
            self = .decodeThroughput(benchmarkId: id, prefillTokens: n[0], decodeTokens: n[1])
        } else if let n = suffixInts(after: "end_to_end_latency"), n.count == 2 {
            self = .endToEndLatency(benchmarkId: id, prefillTokens: n[0], decodeTokens: n[1])
        } else {
            return nil
        }
    }
}

extension BenchmarkDefinition {
    /// The typed benchmark kind for this definition — the canonical tag that lets
    /// the two enums stop drifting via a hand-written string bridge. `BenchmarkType`
    /// is the single source of truth for the `benchmark_type` vocabulary; a parsed
    /// definition maps to its case here, the inverse of the `init(from:)` tag switch.
    nonisolated var type: BenchmarkType {
        switch self {
        case .prefillThroughput: return .prefillThroughput
        case .decodeThroughput: return .decodeThroughput
        case .endToEndLatency: return .endToEndLatency
        case .maxMemoryUsage: return .maxMemoryUsage
        case .eval: return .eval
        case .vlThroughput: return .vlThroughput
        }
    }

    /// The `benchmark_type` wire string — `type.rawValue`, kept as a named
    /// accessor for the string-keyed catalog / CSV readers that still index by it.
    nonisolated var benchmarkType: String { type.rawValue }

    /// A minimal `parameter_*` dict reconstructed from the typed params — the shape the
    /// string-keyed readers index (the live cell label, CSV export), so a parsed-only cell
    /// presents the same fields a catalog-backed one does.
    nonisolated var parameterFields: [String: Any] {
        switch self {
        case .prefillThroughput(_, let p), .maxMemoryUsage(_, let p):
            return ["parameter_prefill_tokens": Int(p)]
        case .decodeThroughput(_, let p, let d), .endToEndLatency(_, let p, let d):
            return ["parameter_prefill_tokens": Int(p), "parameter_decode_tokens": Int(d)]
        case .vlThroughput(_, let w, let h, let t, let d):
            return [
                "parameter_image_width": Int(w), "parameter_image_height": Int(h),
                "parameter_text_tokens": Int(t), "parameter_decode_tokens": Int(d),
            ]
        case .eval(_, _, _, let maxTokens, _, _):
            return ["parameter_max_tokens": Int(maxTokens)]
        }
    }
}

/// One eval dataset row: a stable `id` and the chat `messages` to complete.
struct EvalSample: Codable, Equatable {
    let id: String
    let messages: [[String: String]]
}

nonisolated extension BenchmarkDefinition: Decodable {
    /// The flat wire fields (`benchmark_type` tag + `parameter_*` params), matching
    /// the Rust serde shape so the same JSON deserializes on both sides.
    private enum Key: String, CodingKey {
        case benchmarkId = "benchmark_id"
        case benchmarkType = "benchmark_type"
        case prefillTokens = "parameter_prefill_tokens"
        case decodeTokens = "parameter_decode_tokens"
        case evalId = "parameter_eval_id"
        case datasetName = "parameter_dataset_name"
        case maxTokens = "parameter_max_tokens"
        case mcqChoices = "parameter_mcq_choices"
        case samples
        case imageWidth = "parameter_image_width"
        case imageHeight = "parameter_image_height"
        case textTokens = "parameter_text_tokens"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: Key.self)
        let id = try c.decode(String.self, forKey: .benchmarkId)
        // Decode the tag as the enum — an unknown `benchmark_type` throws via the
        // synthesized rawValue decode (DecodingError), preserving reject-unknown.
        // Switching over `BenchmarkType` is exhaustive, so it can't drift from the
        // enum's cases (no literal-string switch, no `default:`).
        let type = try c.decode(BenchmarkType.self, forKey: .benchmarkType)
        switch type {
        case .prefillThroughput:
            self = .prefillThroughput(
                benchmarkId: id, prefillTokens: try c.decode(UInt32.self, forKey: .prefillTokens))
        case .decodeThroughput:
            self = .decodeThroughput(
                benchmarkId: id,
                prefillTokens: try c.decode(UInt32.self, forKey: .prefillTokens),
                decodeTokens: try c.decode(UInt32.self, forKey: .decodeTokens))
        case .endToEndLatency:
            self = .endToEndLatency(
                benchmarkId: id,
                prefillTokens: try c.decode(UInt32.self, forKey: .prefillTokens),
                decodeTokens: try c.decode(UInt32.self, forKey: .decodeTokens))
        case .maxMemoryUsage:
            self = .maxMemoryUsage(
                benchmarkId: id, prefillTokens: try c.decode(UInt32.self, forKey: .prefillTokens))
        case .eval:
            self = .eval(
                benchmarkId: id,
                evalId: try c.decode(EvalId.self, forKey: .evalId),
                datasetName: try c.decode(String.self, forKey: .datasetName),
                maxTokens: try c.decode(UInt32.self, forKey: .maxTokens),
                mcqChoices: try c.decodeIfPresent([String].self, forKey: .mcqChoices),
                samples: try c.decodeIfPresent([EvalSample].self, forKey: .samples))
        case .vlThroughput:
            self = .vlThroughput(
                benchmarkId: id,
                imageWidth: try c.decode(UInt32.self, forKey: .imageWidth),
                imageHeight: try c.decode(UInt32.self, forKey: .imageHeight),
                textTokens: try c.decode(UInt32.self, forKey: .textTokens),
                decodeTokens: try c.decode(UInt32.self, forKey: .decodeTokens))
        }
    }

    /// The `benchmark_type` tags this client understands — derived from
    /// `BenchmarkType` so there's no parallel string list to drift (one source of
    /// truth, mirroring Rust `KnownEvalId`). Lets a consumer tell an unrecognized
    /// type (skip quietly) from a known type that failed to decode (a schema
    /// mismatch worth logging).
    static let knownTypes: Set<String> = Set(BenchmarkType.allCases.map(\.rawValue))
}

nonisolated extension BenchmarkDefinition: Encodable {
    /// The inverse of `init(from:)`, emitting the same flat tag + `parameter_*` shape.
    ///
    /// Needed because the local catalog half is *written* by this client — the crate
    /// serializes the same `BenchmarkDefinition` into `benchmarks/local/<id>.json`. Each
    /// case writes exactly the keys its decode requires, so a definition round-trips;
    /// `BenchmarkDefinitionCodingTests` pins that.
    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: Key.self)
        try c.encode(benchmarkId, forKey: .benchmarkId)
        try c.encode(type, forKey: .benchmarkType)
        switch self {
        case .prefillThroughput(_, let prefill), .maxMemoryUsage(_, let prefill):
            try c.encode(prefill, forKey: .prefillTokens)
        case .decodeThroughput(_, let prefill, let decode),
             .endToEndLatency(_, let prefill, let decode):
            try c.encode(prefill, forKey: .prefillTokens)
            try c.encode(decode, forKey: .decodeTokens)
        case .eval(_, let evalId, let dataset, let maxTokens, let mcqChoices, let samples):
            try c.encode(evalId, forKey: .evalId)
            try c.encode(dataset, forKey: .datasetName)
            try c.encode(maxTokens, forKey: .maxTokens)
            // Absent, not null — the decode side uses `decodeIfPresent`.
            try c.encodeIfPresent(mcqChoices, forKey: .mcqChoices)
            try c.encodeIfPresent(samples, forKey: .samples)
        case .vlThroughput(_, let width, let height, let textTokens, let decode):
            try c.encode(width, forKey: .imageWidth)
            try c.encode(height, forKey: .imageHeight)
            try c.encode(textTokens, forKey: .textTokens)
            try c.encode(decode, forKey: .decodeTokens)
        }
    }
}

File: benchmark-tablets/pipette-clients/android/Pipette/app/src/main/java/ai/liquid/pipette/BenchmarkDefinition.kt
package ai.liquid.pipette

import org.json.JSONArray
import org.json.JSONObject

/**
 * Benchmark kinds, mirroring pipette-ops `BenchmarkType`. `wire` is the `benchmark_type` tag used in the JSON the engine and management server
 * exchange; it matches the Rust serde representation exactly.
 */
enum class BenchmarkType(val wire: String, val displayName: String, val rank: Int) {
  END_TO_END_LATENCY("end_to_end_latency", "End-to-End Latency", 0),
  PREFILL_THROUGHPUT("prefill_throughput", "Prefill Throughput", 1),
  DECODE_THROUGHPUT("decode_throughput", "Decode Throughput", 2),
  MAX_MEMORY_USAGE("max_memory_usage", "Max Memory Usage", 3),
  VL_THROUGHPUT("vl_throughput", "Vision-Language Throughput", 4),
  EVAL("eval", "Eval Accuracy", 5);

  /**
   * Whether this benchmark must observe the model load itself, so the runner has to load fresh inside the measured window rather than reuse a
   * resident handle. Only [MAX_MEMORY_USAGE], whose whole measurement is the peak the load produces.
   *
   * Lives on the type, mirroring iOS `BenchmarkType.requiresFreshLoad`, because a call site that restates the rule is free to restate it wrongly:
   * this began as `cell.benchmarkType == MAX_MEMORY_USAGE.wire` in [JobRunner], which compares a nullable wire string and quietly answered false for
   * a cell whose manifest predates the stored type.
   */
  val requiresFreshLoad: Boolean
    get() = this == MAX_MEMORY_USAGE

  companion object {
    fun fromWire(wire: String): BenchmarkType? = entries.firstOrNull { it.wire == wire }
  }
}

/**
 * Typed benchmark definition, mirroring pipette-ops `BenchmarkDefinition` (a `benchmark_type`-tagged enum with per-kind parameters). [toJson]
 * serializes to the same flat, tagged shape the native engine and the management server consume.
 */
sealed class BenchmarkDefinition {
  abstract val benchmarkId: BenchmarkId
  abstract val type: BenchmarkType

  /** Wire `benchmark_type` tag — convenience for the String-based call sites. */
  val benchmarkType: String
    get() = type.wire

  /** The flat, `benchmark_type`-tagged JSON the engine/server expect. */
  abstract fun toJson(): JSONObject

  /** Alias for [toJson]; the engine bridge is fed this JSON. */
  val rawJson: JSONObject
    get() = toJson()

  /** Short human label for the parameters (used in the picker/CSV). */
  abstract val label: String

  protected fun base(): JSONObject = JSONObject().put("benchmark_id", benchmarkId.toString()).put("benchmark_type", type.wire)

  data class PrefillThroughput(override val benchmarkId: BenchmarkId, val prefillTokens: Int) : BenchmarkDefinition() {
    override val type = BenchmarkType.PREFILL_THROUGHPUT

    override fun toJson(): JSONObject = base().put("parameter_prefill_tokens", prefillTokens)

    override val label: String
      get() = "${prefillTokens}tok in"
  }

  data class DecodeThroughput(override val benchmarkId: BenchmarkId, val prefillTokens: Int, val decodeTokens: Int) : BenchmarkDefinition() {
    override val type = BenchmarkType.DECODE_THROUGHPUT

    override fun toJson(): JSONObject = base().put("parameter_prefill_tokens", prefillTokens).put("parameter_decode_tokens", decodeTokens)

    override val label: String
      get() = "${prefillTokens}tok in - $decodeTokens tok out"
  }

  data class EndToEndLatency(override val benchmarkId: BenchmarkId, val prefillTokens: Int, val decodeTokens: Int) : BenchmarkDefinition() {
    override val type = BenchmarkType.END_TO_END_LATENCY

    override fun toJson(): JSONObject = base().put("parameter_prefill_tokens", prefillTokens).put("parameter_decode_tokens", decodeTokens)

    override val label: String
      get() = "${prefillTokens}tok in - $decodeTokens tok out"
  }

  data class MaxMemoryUsage(override val benchmarkId: BenchmarkId, val prefillTokens: Int) : BenchmarkDefinition() {
    override val type = BenchmarkType.MAX_MEMORY_USAGE

    override fun toJson(): JSONObject = base().put("parameter_prefill_tokens", prefillTokens)

    override val label: String
      get() = "${prefillTokens}tok context"
  }

  data class VlThroughput(
    override val benchmarkId: BenchmarkId,
    val imageWidth: Int,
    val imageHeight: Int,
    val textTokens: Int,
    val decodeTokens: Int,
  ) : BenchmarkDefinition() {
    override val type = BenchmarkType.VL_THROUGHPUT

    override fun toJson(): JSONObject =
      base()
        .put("parameter_image_width", imageWidth)
        .put("parameter_image_height", imageHeight)
        .put("parameter_text_tokens", textTokens)
        .put("parameter_decode_tokens", decodeTokens)

    override val label: String
      get() = "${imageWidth}x$imageHeight - ${textTokens}tok text - $decodeTokens tok out"
  }

  data class Eval(
    override val benchmarkId: BenchmarkId,
    val evalId: String,
    val datasetName: String,
    val maxTokens: Int,
    val mcqChoices: List<String>?,
    val samples: JSONArray?,
  ) : BenchmarkDefinition() {
    override val type = BenchmarkType.EVAL

    override fun toJson(): JSONObject {
      val json = base().put("parameter_eval_id", evalId).put("parameter_dataset_name", datasetName).put("parameter_max_tokens", maxTokens)
      mcqChoices?.let { json.put("parameter_mcq_choices", JSONArray(it)) }
      samples?.let { json.put("samples", it) }
      return json
    }

    override val label: String
      get() = benchmarkId.toString()
  }

  /**
   * The outcome of reading one server catalog entry, split so a consumer can apply the tolerance policy PIP-248 specifies: an unrecognized type is
   * skipped quietly, a known type that fails to decode is skipped and logged as an error. Mirrors what iOS gets from a thrown `DecodingError` plus
   * [BenchmarkType] membership.
   */
  sealed interface ParseResult {
    data class Ok(val definition: BenchmarkDefinition) : ParseResult

    /**
     * `benchmark_type` is not in this client's enum. Quiet on purpose: a newer server catalog reaching an older client is expected, not a fault, and
     * logging it as one trains readers to ignore the log.
     */
    data class UnknownType(val benchmarkId: String, val type: String) : ParseResult

    /**
     * A type we understand, carrying parameters we cannot read. Worth an error: it means the server and this client disagree about a shape they are
     * both supposed to implement, which is a bug on one side rather than a version skew.
     */
    data class SchemaMismatch(val benchmarkId: String, val type: String, val detail: String) : ParseResult
  }

  companion object {
    /** Parse the flat, `benchmark_type`-tagged JSON into a typed definition, or null for any entry this client cannot fully read. */
    fun fromJson(obj: JSONObject): BenchmarkDefinition? = (parse(obj) as? ParseResult.Ok)?.definition

    /**
     * [fromJson] with the reason attached, for the sync's tolerance policy.
     *
     * Strict about required parameters, deliberately. These read with `getInt`/`getString`, which throw on a missing or ill-typed key, rather than
     * the `optInt`/`optString` this used to use, which return `0`/`""`. Defaulting is the wrong kind of tolerance here: a `decode_throughput` whose
     * `parameter_decode_tokens` is absent would decode to zero tokens and then *run*, producing a measurement of nothing that no consumer can tell
     * from a real one. Skipping it is recoverable; a plausible wrong number is not. iOS (`try c.decode(UInt32.self, …)`) and the Rust client
     * (`serde_json::from_value`) have always required these, so this is Android converging on the other two rather than a new rule.
     *
     * `parameter_mcq_choices` and `samples` stay optional, matching iOS's `decodeIfPresent`: only the per-id detail response carries `samples`, so
     * requiring it would reject every entry the list endpoint returns.
     */
    fun parse(obj: JSONObject): ParseResult {
      val rawId = obj.optString("benchmark_id").ifEmpty { "<no id>" }
      val rawType = obj.optString("benchmark_type").ifEmpty { "nil" }
      // Classified by type first so a bad id on a known type reads as a schema mismatch rather than
      // as an unknown type, matching how iOS's logSkip classifies a thrown decode.
      val type = BenchmarkType.fromWire(obj.optString("benchmark_type")) ?: return ParseResult.UnknownType(rawId, rawType)
      return runCatching { ParseResult.Ok(decodeKnown(obj, type)) }
        .getOrElse { error -> ParseResult.SchemaMismatch(rawId, rawType, error.message ?: error.javaClass.simpleName) }
    }

    /** Throws (`JSONException`, or [PlanTypeError] via [BenchmarkId]) when the entry doesn't carry what [type] requires. */
    private fun decodeKnown(obj: JSONObject, type: BenchmarkType): BenchmarkDefinition {
      val id = BenchmarkId.parse(obj.getString("benchmark_id"))
      return when (type) {
        BenchmarkType.PREFILL_THROUGHPUT -> PrefillThroughput(id, obj.getInt("parameter_prefill_tokens"))
        BenchmarkType.DECODE_THROUGHPUT -> DecodeThroughput(id, obj.getInt("parameter_prefill_tokens"), obj.getInt("parameter_decode_tokens"))
        BenchmarkType.END_TO_END_LATENCY -> EndToEndLatency(id, obj.getInt("parameter_prefill_tokens"), obj.getInt("parameter_decode_tokens"))
        BenchmarkType.MAX_MEMORY_USAGE -> MaxMemoryUsage(id, obj.getInt("parameter_prefill_tokens"))
        BenchmarkType.VL_THROUGHPUT ->
          VlThroughput(
            id,
            obj.getInt("parameter_image_width"),
            obj.getInt("parameter_image_height"),
            obj.getInt("parameter_text_tokens"),
            obj.getInt("parameter_decode_tokens"),
          )
        BenchmarkType.EVAL ->
          Eval(
            id,
            obj.getString("parameter_eval_id"),
            obj.getString("parameter_dataset_name"),
            obj.getInt("parameter_max_tokens"),
            obj.optJSONArray("parameter_mcq_choices")?.let { arr -> List(arr.length()) { arr.getString(it) } },
            obj.optJSONArray("samples"),
          )
      }
    }

    /**
     * Reconstruct a definition from a structured benchmark id, for the four ladder types — the resolution fallback when the synced catalog has no
     * entry for an id (e.g. a historical job whose benchmark left the catalog, or before the first sync). Mirrors iOS
     * `BenchmarkDefinition(parsingId:)` and the Rust `BenchmarkType::from_id`: the type is the matched prefix and the workload numbers are parsed
     * straight out of the id.
     * - `prefill_throughput_<P>` → [PrefillThroughput]
     * - `max_memory_usage_<P>` → [MaxMemoryUsage]
     * - `decode_throughput_<P>_<D>` → [DecodeThroughput]
     * - `end_to_end_latency_<P>_<D>` → [EndToEndLatency]
     *
     * Returns null for `eval`, `vl_throughput`, `_smoke`, or anything that doesn't match a known type with the expected trailing integers.
     */
    fun parseId(id: String): BenchmarkDefinition? {
      val bid = BenchmarkId.parseOrNull(id) ?: return null
      // (type prefix, arity, builder over the trailing integers). First match whose id has exactly `arity` integer segments wins.
      val ladder =
        listOf<Triple<String, Int, (List<Int>) -> BenchmarkDefinition>>(
          Triple("prefill_throughput", 1) { PrefillThroughput(bid, it[0]) },
          Triple("max_memory_usage", 1) { MaxMemoryUsage(bid, it[0]) },
          Triple("decode_throughput", 2) { DecodeThroughput(bid, it[0], it[1]) },
          Triple("end_to_end_latency", 2) { EndToEndLatency(bid, it[0], it[1]) },
        )
      return ladder.firstNotNullOfOrNull { (type, arity, build) -> suffixInts(id, type)?.takeIf { it.size == arity }?.let(build) }
    }

    /** The trailing integers after `<type>_` in [id], or null if the prefix doesn't match or any segment isn't an integer. */
    private fun suffixInts(id: String, type: String): List<Int>? {
      val prefix = "${type}_"
      if (!id.startsWith(prefix)) return null
      val parts = id.removePrefix(prefix).split("_")
      val nums = parts.mapNotNull { it.toIntOrNull() }
      return if (parts.isNotEmpty() && nums.size == parts.size) nums else null
    }
  }
}
